---
title: "Signature Identification"
author: ["Shixiang Wang", "Huimin Li", "Xuan Wang", "Minfang Song", "Xue-Song Liu (Corresponding author)"]
date: "`r Sys.Date()`"
output:
  rmdformats::readthedown:
    highlight: kate
    self_contained: false
    mathjax: true
    lightbox: true
    gallery: true
    toc: 3
bibliography: ref.bib
link-citations: yes
---

```{r knitr_init_02, echo=FALSE}
library(knitr)
library(rmdformats)
## Global options
options(max.print = "75")
opts_chunk$set(
  echo = TRUE,
  comment = "#>",
  collapse = TRUE,
  cache = TRUE,
  dpi = 300,
  warning = FALSE,
  message = FALSE,
  tidy = "styler"
)
opts_knit$set(width = 75)
```

This part has been divided into 4 sections as the following, it describes how to get the
genomic variation (or instability) signatures from mutation profile (including copy number profile
and single base substitution (SBS) profile).

- [Tally variation records](#tally-variation-records)
- [Estimate signature number](#estimate-signature-number)
- [Extract signatures](#extract-signatures)
- [Show signature profile and aetiology](#show-signature-profile-and-aetiology)

All the processes have been implemented in R package [**sigminer**](https://github.com/ShixiangWang/sigminer).

## Tally variation records

### Methods

For SBS profile, same as previously reported [@alexandrov2013signatures], for each sample, we firstly classified mutation records into six substitution subtypes: C>A, C>G, C>T, T>A, T>C, and T>G (all substitutions are referred to by the pyrimidine of the mutated Watson–Crick base pair). Further, each of the substitutions was examined by incorporating information on the bases immediately 5’ and 3’ to each mutated base generating 96 possible mutation types (6 types of substitution ∗ 4 types of 5’ base ∗ 4 types of 3’ base). Each of 96 mutation types is called **component** here.

For copy number profile, we firstly computed the genome-wide distributions of 8 fundamental copy number features for each sample: 

- the breakpoint count per 10 Mb (named **BP10MB**)
- the breakpoint count per chromosome arm (named **BPArm**)
- the copy number of the segments (named **CN**)
- the difference in copy number between adjacent segments (named **CNCP**)
- the lengths of oscillating copy number segment chains (named **OsCN**)
- the log10 based size of segments (named **SS**)
- the minimal number of chromosome with 50% copy number variation (named **NC50**)
- the burden of chromosome (named **BoChr**)

These features were selected as hallmarks of previously reported genomic aberations like chromothripsis or to denote the distribution pattern of copy number events. The former 6 features have been used in @macintyre2018copy to uncover the mutational processes in ovarian carcinoma.

Next, unlike @macintyre2018copy applied mixture modeling to separate the first 6 copy number features distributions into mixtures of Poisson or Gaussian distributions, we directly classified 8 copy number features distributions into different components according to
the comprehensive consideration of value range, abundance and biological significance. Most of the result are discrete values, and the others are range values.

The setting of 8 features with `r nrow(sigminer::CN.features)` components are shown as below.

```{r}
DT::datatable(sigminer::CN.features)
```

> When component is a discrete value (label is 'point'), min=max.
> 
> When component is a range value (label is 'range'), the range is left open and right closed.
>
> Of note, the blank in `min` column is `-Inf` and the blank in `max` column is `Inf`.
> `[23]` in `BoChr` represents chromosome X (chromosome Y is excluded).

Compare to the method from @macintyre2018copy, our method has several advantages:

1. the computation is more efficient.
2. the meaning of component is easier to read and understand.
3. the new features **NC50** and **BoChr** can be used to determine the distribution pattern of copy number events (global or local).
and the contribution of each chromosome.
4. most importantly, the classification is fixed, so it is much easier to compare the signatures within/across tumor types, the results across different studies and the result signatures to reference signatures. In a word, our method constructs the standard to extract, study and compare copy number signatures.
5. our method is more extensible.

I will talk more about No.1-4 in an individual section (**TBD**), the No.5 will be described in [sigminer vignette]()(**TBD**).

We will generate sample-by-component matrix using the methods above and treate the result as the input of nonnegative matrix decomposition (NMF) algorithm for extracting signatures.

### Tally records

#### Copy number

Read absolute copy number profile as a `CopyNumber` object.

```{r}
# Load packages -----------------------------------------------------------

suppressPackageStartupMessages(library(sigminer))
suppressPackageStartupMessages(library(tidyverse))
suppressPackageStartupMessages(library(NMF))
```


```{r, eval=FALSE}
# Set this per R session
options(sigminer.sex = "male", sigminer.copynumber.max = 20L)

# Generate CopyNumber object ----------------------------------------------

CNV.seqz <- read_copynumber("../data/CNV_from_sequenza.tsv",
  genome_build = "hg38",
  complement = FALSE, verbose = TRUE
)

# remove WCMC160-SRR3146971 with only one CNV
CNV.seqz <- subset(CNV.seqz, subset = !sample %in% "WCMC160-SRR3146971")
save(CNV.seqz, file = "../output/CNV.seqz.RData")
```

Tally the alteration records.

```{r, eval=FALSE}
ncores <- 20 # Set a proper number of thread

# Use classfication method devised by me ("W")
CNV.seqz.tally.W <- sig_tally(CNV.seqz, method = "W", cores = ncores, feature_setting = CN.features)
save(CNV.seqz.tally.W, file = "../output/CNV.seqz.tally.W.RData")
```

#### SBS

Read SBS profile as a `MAF` object.

```{r, eval=FALSE}
# Reading data ------------------------------------------------------------
Maf <- data.table::fread("/public/data/maf/all.maf")
# Remove all NA columns
Maf <- Maf[, which(unlist(lapply(Maf, function(x) !all(is.na(x))))), with = F]
Maf <- read_maf(Maf)
save(Maf, file = "../output/PRAD_TCGA_plus_dbGap_Maf.RData")
```

The result `MAF` object can be used in any analysis provided by R package **maftools** [@mayakonda2016maftools].

Tally the SBS records.

```{r, eval=FALSE}
Maf.tally <- sig_tally(Maf,
                       cores = ncores, ref_genome = "BSgenome.Hsapiens.UCSC.hg38",
                       useSyn = TRUE)
save(Maf.tally, file = "../output/PRAD_TCGA_plus_dbGap_Maf.tally.RData")
```

## Estimate signature number

Here, r package **NMF** [@gaujoux2010flexible] is used for running NMF algorithm
Factorization rank `r` in NMF defines the number of signatures used to approximate the target sample-by-component matrix.
A common way for deciding on `r` is to try different values, compute some quality measure of the results, and choose the best
value according to this quality criteria.

As suggested, performing 30-50 runs is considered sufficient to get a robust estimation of the `r` value. Here we perform 50 runs.

### Copy number

Estimate the number of copy number signature from 2 to 12.

```{r, eval=FALSE}
EST.seqz.W.all <- sig_estimate(CNV.seqz.tally.W$nmf_matrix,
  range = 2:12, nrun = 50, cores = ncores, use_random = TRUE,
  save_plots = FALSE,
  verbose = TRUE
)
save(EST.seqz.W.all, file = "../output/EST.seqz.W.all.RData")
```

### SBS

Estimate the number of SBS signature from 2 to 10.

```{r, eval=FALSE}
# Remove the effect of hyper mutated samples (not removing hyper-mutated samples)
# The idead is adopted from SignatureAnalyzer package
nmf_matrix <- handle_hyper_mutation(Maf.tally$nmf_matrix)

EST.Maf <- sig_estimate(nmf_matrix,
  range = 2:10, nrun = 50, cores = ncores, use_random = TRUE,
  save_plots = FALSE,
  verbose = TRUE
)

save(EST.Maf, file = "../output/EST.PRAD_TCGA_plus_dbGap_Maf.RData")
save(nmf_matrix, file = "../output/Maf_matrix.RData")
```


### Result

Let's load the estimated results.

```{r}
load(file = "../output/EST.seqz.W.all.RData")
load(file = "../output/EST.PRAD_TCGA_plus_dbGap_Maf.RData")
```

The most common approach is to use the cophenetic correlation coefficient. I show the measure vs. signature number as below.

```{r}
show_sig_number_survey(EST.seqz.W.all, right_y = NULL)
```

The best possible signature number value is the one at which the cophenetic correlation value on the y-axis drops significantly [@gaujoux2010flexible]. Here we can take that the value `5` is the optimal signature number. For `3`, it is small to produce the result with biological significance due to the hugo contribution of CNV in prostate cancer; for `8`, it is a little bigger, so it may be not easy to understand the result considering currently we know few about the mechanism and consequence of copy number events.

```{r}
p = show_sig_number_survey(EST.seqz.W.all, right_y = NULL)
add_h_arrow(p, x = 5.1, y = 0.99, seg_len = 0.5)
```

Next we take a look at SBS.


```{r}
p = show_sig_number_survey(EST.Maf, right_y = NULL)
add_h_arrow(p, x = 3.1, y = 0.982, seg_len = 0.5)
```

For SBS, we select `3`. We have the knowledge that `3` mutational signatures from TCGA prostate cancers are recorded in COSMIC database (COSMIC 1, 5, 6), and the cophenetic has a sharp decrease after `3` indicating that substantially less stability is achieved using more than `3` clusters.

## Extract signatures

Now that we have determined the signature number, we then extract the signatures with 50 runs.

Copy number:

```{r, eval=FALSE}
# Extract signatures ------------------------------------------------------
Sig.CNV.seqz.W <- sig_extract(CNV.seqz.tally.W$nmf_matrix, n_sig = 5, nrun = 50, cores = ncores)
save(Sig.CNV.seqz.W, file = "../output/Sig.CNV.seqz.W.RData")
```

SBS:

```{r, eval=FALSE}
# Extract signatures ------------------------------------------------------
Sig.SNV <- sig_extract(nmf_matrix, n_sig = 3, nrun = 50, cores = ncores)
save(Sig.SNV, file = "../output/Sig.PRAD_TCGA_plus_dbGap_Maf.RData")
```

## Show signature profile and aetiology

Now that we have the extracted signatures, we can show the signature profile.

```{r}
load("../output/Sig.PRAD_TCGA_plus_dbGap_Maf.RData")
load("../output/Sig.CNV.seqz.W.RData")
```


### SBS signature profile


```{r, fig.width=10, fig.height=5}
# I provide two style 'default' and 'cosmic' in sigminer
show_sig_profile(Sig.SNV, mode = "SBS", style = "cosmic", x_label_angle = 90, x_label_vjust = 0.5)
```

> SBS signatures are displayed based on the observed component frequency of the human genome, i.e., representing the relative proportions of mutations generated by each signature based on the actual trinucleotide frequencies of the reference human genome.

The aetiology of SBS signatures is more clear than the aetiology of copy number signatures. We can find aetiologies of the 3 signatures above by computing their cosine similarity to COSMIC reference signatures.

We compare the 3 signature to COSMIC signature database v2 and v3.

```{r, message=TRUE}
sim_v2 = get_sig_similarity(Sig.SNV, sig_db = "legacy")
```

```{r, message=TRUE}
sim_v3 = get_sig_similarity(Sig.SNV, sig_db = "SBS")
```

From the results above, basically we can confirm that the aetiology of `Sig2` is **dMMR** (defects in mismatch repair) and the aetiology of `Sig3` is **aging**. The aetiology of `Sig1` may be **HRD** (Defective homologous recombination-based DNA damage repair) or an unknown reason.

```{r, fig.width=10, fig.height=5}
p = show_sig_profile(Sig.SNV, mode = "SBS", style = "cosmic", x_label_angle = 90, x_label_vjust = 0.5)
add_labels(p, x = 0.92, y = 0.3, y_end = 0.85, n_label = 3, 
           labels = rev(c("HRD or unknown", "dMMR", "Aging")), hjust = 1)
```

The exposure of signatures in each sample is plotted as the below.

```{r}
show_sig_exposure(Sig.SNV, rm_space = TRUE, style = "cosmic")
```

Set a cutoff 2000 to remove some outliers.

```{r}
show_sig_exposure(Sig.SNV, 
                  cutoff = 2000,
                  rm_space = TRUE, style = "cosmic")
```

> The `Est_Counts` panel shows the estimated SBS count.

We can clearly see that 

- most of samples have very few SBS signature exposures.
- `Sig1` and `Sig3` are the major resources.

### Copy number signature profile

```{r, fig.width=12, fig.height=7}
show_sig_profile(Sig.CNV.seqz.W, style = "cosmic", method = "W", normalize = "feature")
```

> Similar to SBS signatures, copy number signatures are displayed based on the observed component frequency of the human genome.
> Of note, considering the count process of each feature is relatively independent, the profile is row normalized by each feature, 
> unlike @macintyre2018copy did column normalization (this method is easy to mislead readers),
> so the bar height can be compared within/between features.

There is no reference database for copy number signatures, we cannot get aetiologies of the 5 signatures by similarity computation. By reading the description of previous reports [@macintyre2018copy;@yi2018patterns] and studying the signature profile carefully, we determine or limite aetiologies of the 5 signatures with also the downstream analyses supported.

- Sig1 - Focal amplification due to failure of cell cyle control
  - large copy number changes resulting high copy number states
- Sig2 - HRD;MMBIR;Tandem duplication through CDK12 inactivation
  - match the pattern of MMBIR (microhomology mediated break-induced replication): many long **OsCN** with copy gains
  - related to CDK12 mutation
  - high number of breakpoints
  - small segment size
  - related to COSMIC 5 (v2) (from @macintyre2018copy)
- Sig3 - Whole-genome duplication (due to failure of cell cycle control)
  - high segment copy number
  - few OsCN
  - global CNV distribution
- Sig4 - Chromothripsis
  - 1 copy number deletions resulting considerable small OsCN
  - large segment size
- Sig5 -  BFB and copy number neutral like events including Chromoplexy, translocation, and etc. (these may be early driver events)
  - few CNV with focal distribution
  - positively associated with gene fusion

The exposure of signatures in each sample is plotted as the below.

```{r}
show_sig_exposure(Sig.CNV.seqz.W, rm_space = TRUE, style = "cosmic")
```

> The `Est_Counts` panel shows the estimated copy number segment count.

### Show similarity matrix

```{r, fig.width=11, fig.height=6}
pheatmap::pheatmap(sim_v2$similarity,
  cluster_rows = F
)
```

```{r, fig.width=15, fig.height=6}
pheatmap::pheatmap(sim_v3$similarity,
  cluster_rows = F
)
```

### Check the signature exposure

Signature exposure is the estimation of mutation calling records. Here we check this by correlation analysis.

#### Mutation

Observe total SBS mutations from data. 

```{r}
load(file = "../output/PRAD_TCGA_plus_dbGap_Maf.RData")
Maf_dt = rbind(Maf@data, Maf@maf.silent)
Mut_dt = Maf_dt[Variant_Type == "SNP", .(Total = .N), by = Tumor_Sample_Barcode]
```

Estimate SBS mutations from signature exposure.

```{r}
Mut_expo = get_sig_exposure(Sig.SNV)[, .(sample, Est=Sig1+Sig2+Sig3)]
```

Merge the two `data.table` above.

```{r}
Mut_merged_dt = merge(Mut_dt, Mut_expo, by.x = "Tumor_Sample_Barcode", by.y = "sample")
# log10 the data
Mut_merged_dt[, `:=`(Total = log10(Total), Est = log10(Est))]

head(Mut_merged_dt)
```

Now we can see their correlation.

```{r}
ggpubr::ggscatter(Mut_merged_dt, x = "Total", y = "Est", add = "reg.line",
                  xlab = "Log10 based total SBS counts (observed)",
                  ylab = "Log10 based total SBS counts (estimated)") +
  ggpubr::stat_cor(label.x.npc = 0.5, label.y.npc = 0.8)
```

We can know that we did a perfect NMF for SBS data.

#### Copy number

Observe total segment counts from copy number data. 

```{r}
load(file = "../output/CNV.seqz.RData")
CN_dt = CNV.seqz@data[, .(Total = .N), by = sample]
```

Estimate segment counts from signature exposure.

```{r}
CN_expo = get_sig_exposure(Sig.CNV.seqz.W)[, .(sample, Est=Sig1+Sig2+Sig3+Sig4+Sig5)]
```

Merge the two `data.table` above.

```{r}
CN_merged_dt = merge(CN_dt, CN_expo, by = "sample")

head(CN_merged_dt)
```


Now we can see their correlation.

```{r}
ggpubr::ggscatter(CN_merged_dt, x = "Total", y = "Est", add = "reg.line",
                  xlab = "Log10 based total segment counts (observed)",
                  ylab = "Log10 based total segment counts (estimated)") +
  ggpubr::stat_cor(label.x.npc = 0.5, label.y.npc = 0.8, hjust = 1)
```

We can see that the estimation of copy number segment count fits observed data well. The residue can be further visualized by histgram.

```{r}
CN_merged_dt[, Residue:=Est-Total]
```

```{r}
library(ggplot2)
ggplot(CN_merged_dt) +
  geom_histogram(aes(x = Residue))
```


***
