---
title: "Association Analysis"
author: ["Shixiang Wang", "Huimin Li", "Xuan Wang", "Minfang Song", "Xue-Song Liu (Corresponding author)"]
date: "`r Sys.Date()`"
output:
  rmdformats::readthedown:
    highlight: kate
    self_contained: false
    mathjax: true
    lightbox: true
    gallery: true
    toc: 3
bibliography: ref.bib
link-citations: yes
---

```{r knitr_init_03, echo=FALSE}
library(knitr)
library(rmdformats)
## Global options
options(max.print = "75")
opts_chunk$set(
  echo = TRUE,
  comment = "#>",
  collapse = TRUE,
  cache = TRUE,
  dpi = 300,
  warning = FALSE,
  message = TRUE,
  tidy = "styler"
)
opts_knit$set(width = 75)
```

This part has been divided into 2 sections: data integration and association analysis.

## Data integration

In this section, I integrate all genotype/phenotype data as a tidy data table used for downstream analyses.

Load packages and prepared data.

```{r, message=FALSE}
# Integrate all informaiton to sample level
library(tidyverse)
library(sigminer)
library(maftools)

# Loading clinical related data -------------------------------------------

Info <- readRDS("../data/PRAD_CLINICAL.rds")
# Purity and ploidy info from sequenza
PurityInfo <- read_tsv("../data/PRAD_Purity_and_Ploidy_Sequenza.tsv")

# Processing CNV data -----------------------------------------------------

load("../output/CNV.seqz.RData")
load("../output/Sig.CNV.seqz.W.RData")
CNV <- CNV.seqz
Sig.CNV <- Sig.CNV.seqz.W
rm(Sig.CNV.seqz.W, CNV.seqz)
```

After running NMF, we can get robust clusters from consensus matrix, more see `?NMF::predict`. 

I also create functions `get_sig_exposure()` to get the absolute/relative exposure of signatures and `scoring()` to
get scores of typical copy number features. More run `?sigminer::get_sig_exposure` and `?sigminer::scoring` in your
R console or see the package manual.

> Relative exposure is more useful for clustering and absolute exposure is more useful for association analysis.

```{r}
# CNV
CNVGroupInfo <- get_groups(Sig.CNV, method = "consensus", match_consensus = TRUE)
CNVInfo <- CNV@summary.per.sample
CNVExposureInfo <- get_sig_exposure(Sig.CNV)
CNVscores <- scoring(CNV)
```

We find that the group `1` is assigned to `Sig3` due to sample with `Sig3` dominant has the maximum fraction in group `1`. However, group `1` is the only group with `Sig1` enriched. So we modify this result.

```{r}
CNVGroupInfo$enrich_sig[CNVGroupInfo$group == "1"] = "Sig1"
```

Next, we process mutation data to get TMB, driver info, clusters, and etc (detail of some step has been described in [R script](https://github.com/ShixiangWang/prad_signature/blob/master/analysis/src/03-data-integration-updates.R) of the repo).

```{r}
# Processing mutation data ------------------------------------------------

load(file = "../output/PRAD_TCGA_plus_dbGap_Maf.RData")
load(file = "../output/Sig.PRAD_TCGA_plus_dbGap_Maf.RData")

Maf_samp = data.table::data.table(
  Tumor_Sample_Barcode = unique(rbind(Maf@data, Maf@maf.silent)$Tumor_Sample_Barcode)
)

TMBInfo <- Maf@variant.type.summary
TMBInfo$n_INDEL = TMBInfo$DEL + TMBInfo$INS
TMBInfo = TMBInfo[, .(Tumor_Sample_Barcode, n_INDEL, SNP)]
colnames(TMBInfo)[3] = "n_SNV"
TMBInfo = merge(Maf_samp, TMBInfo, by = "Tumor_Sample_Barcode", all = TRUE)
# Fill NAs
TMBInfo = TMBInfo %>%
  dtplyr::lazy_dt() %>%
  dplyr::mutate_at(vars(dplyr::starts_with("n_")), ~ifelse(is.na(.), 0, .)) %>%
  data.table::as.data.table()

load(file = "../output/PRAD_driver_info.RData")
load(file = "../output/PRAD_heter_info.RData")

SNVGroupInfo <- get_groups(Sig.SNV, method = "consensus", match_consensus = TRUE)
SNVExposureInfo <- get_sig_exposure(Sig.SNV)

# Processing gene and pathway mutation ------------------------------------

load(file = "../output/PRAD_gene_and_pathway_mutation.RData")
```

Next, we keep keys of all `data.frame`s are same and merge them into one.


```{r}
# Merge data --------------------------------------------------------------
Info <- Info %>%
  dplyr::mutate(
    CNV_ID = dplyr::case_when(
      !startsWith(tumor_Run, "TCGA") & !is.na(tumor_Run) ~ paste(subject_id, tumor_Run, sep = "-"),
      startsWith(tumor_Run, "TCGA") & !is.na(tumor_Run) ~ tumor_Run,
      TRUE ~ NA_character_
    )
  )

PurityInfo
colnames(CNVGroupInfo) <- c("sample", "cnv_group", "cnv_weight", "cnv_enrich_sig")
CNVInfo

colnames(CNVExposureInfo) <- c("sample", paste0("CN-", colnames(CNVExposureInfo)[-1]))
colnames(TMBInfo)[1] <- "sample"
TMBInfo[, sample := as.character(sample)][, sample := ifelse(startsWith(sample, "TCGA"),
                                                             substr(sample, 1, 15),
                                                             sample
)]

colnames(DriverDF)[1] = "sample"
DriverDF[, sample := as.character(sample)][, sample := ifelse(startsWith(sample, "TCGA"),
                                                              substr(sample, 1, 15),
                                                              sample
)]
colnames(SNVGroupInfo) <- c("sample", "snv_group", "snv_weight", "snv_enrich_sig")
SNVGroupInfo[, sample := as.character(sample)][, sample := ifelse(startsWith(sample, "TCGA"),
                                                                  substr(sample, 1, 15),
                                                                  sample
)]
colnames(SNVExposureInfo) <- c("sample", paste0("SBS-", colnames(SNVExposureInfo)[-1]))
SNVExposureInfo[, sample := as.character(sample)][, sample := ifelse(startsWith(sample, "TCGA"),
                                                                     substr(sample, 1, 15),
                                                                     sample
)]
colnames(TitvInfo) <- c("sample", "Ti_fraction", "Tv_fraction")
TitvInfo[, sample := as.character(sample)][, sample := ifelse(startsWith(sample, "TCGA"),
                                                              substr(sample, 1, 15),
                                                              sample
)]
colnames(MathDF) <- c("sample", "MATH")
MathDF[, sample := as.character(sample)][, sample := ifelse(startsWith(sample, "TCGA"),
                                                            substr(sample, 1, 15),
                                                            sample
)]
colnames(ClusterDF) <- c("sample", "n_mutation_cluster")
ClusterDF[, sample := as.character(sample)][, sample := ifelse(startsWith(sample, "TCGA"),
                                                               substr(sample, 1, 15),
                                                               sample
)]

# data.table::setDT(summary_mutation)
# data.table::setDT(summary_pathway)

colnames(summary_mutation)[1] <- "sample"
colnames(summary_pathway)[1] <- "sample"
summary_mutation[, sample := as.character(sample)][, sample := ifelse(startsWith(sample, "TCGA"),
                                                                      substr(sample, 1, 15),
                                                                      sample
)]
summary_pathway[, sample := as.character(sample)][, sample := ifelse(startsWith(sample, "TCGA"),
                                                                     substr(sample, 1, 15),
                                                                     sample
)]


colnames(CNVInfo)[2:4] = c("n_CNV", "n_Amp", "n_Del")

MergeInfo <- Info[which(!is.na(Info$tumor_Run)), ] %>%
  left_join(CNVInfo[, .(sample, n_CNV, n_Amp, n_Del)], by = c("CNV_ID" = "sample")) %>%
  left_join(CNVscores, by = c("CNV_ID" = "sample")) %>%
  left_join(CNVGroupInfo, by = c("CNV_ID" = "sample")) %>%
  left_join(CNVExposureInfo, by = c("CNV_ID" = "sample")) %>%
  left_join(PurityInfo, by = c("CNV_ID" = "sample")) %>%
  left_join(summary_mutation, by = c("tumor_Run" = "sample")) %>%
  left_join(summary_pathway, by = c("tumor_Run" = "sample")) %>%
  left_join(TMBInfo, by = c("tumor_Run" = "sample")) %>%
  left_join(DriverDF, by = c("tumor_Run" = "sample")) %>%
  dplyr::mutate(
    n_driver = ifelse(!is.na(n_driver), n_driver, 0)
  ) %>%
  left_join(TitvInfo, by = c("tumor_Run" = "sample")) %>%
  left_join(MathDF, by = c("tumor_Run" = "sample")) %>%
  left_join(ClusterDF, by = c("tumor_Run" = "sample")) %>%
  left_join(SNVGroupInfo, by = c("tumor_Run" = "sample")) %>%
  left_join(SNVExposureInfo, by = c("tumor_Run" = "sample")) %>%
  mutate(
    Stage = factor(Stage, ordered = TRUE),
    Fusion = ifelse(Fusion == "Negative", "No", "Yes"),
    sample_type = ifelse(sample_type == "Unknown", "Metastatic", sample_type), # phs000909 are metastatic samples, https://www.ncbi.nlm.nih.gov/projects/gap/cgi-bin/study.cgi?study_id=phs000909.v1.p1
    HasFusion = Fusion,
    HasFusion = ifelse(HasFusion == "Yes", TRUE, FALSE),
    IsMetastatic = ifelse(sample_type == "Metastatic", TRUE, FALSE)
  ) %>% 
  select(-sTDP, -lTDP)


saveRDS(MergeInfo, file = "../output/PRAD_Merge_Info_CNV_from_sequenza_update.RDS")
```

```{r, echo=FALSE}
rm(list = ls())
```


## Association analysis

Now we do association analysis between signatures (via exposure) and genotypes/phenotypes.

### Prepare

Load packages.

```{r, message=FALSE}
library(tidyverse)
library(sigminer)
```

Load the data.

```{r}
df.seqz = readRDS(file = "../output/PRAD_Merge_Info_CNV_from_sequenza_update.RDS")
```

Set the coloumns used for analysis.

```{r}
colnames(df.seqz)[c(19, 23:26)] = c("MACN score", "TD score", "sTD score", "lTD score", "OsCN deletion score")
colnames(df.seqz)[69] = "n_SBS"

# Just use ploidy from sequenza
df.seqz$Ploidy = NULL

# Merge BRCA1 and BRCA2
df.seqz[["BRCA1/2"]] = df.seqz$BRCA1 | df.seqz$BRCA2
# Merge PIK3CA and PIK3CB
df.seqz[["PIK3CA/B"]] = df.seqz$PIK3CA | df.seqz$PIK3CB

# Output for reporting as supp table
openxlsx::write.xlsx(df.seqz, file = "../output/supp_sample_table.xlsx")
# Save for futher use
saveRDS(df.seqz, file = "../output/df.seqz.RDS")

cols_to_sigs.seqz <- c(paste0("CN-Sig", 1:5), paste0("SBS-Sig", 1:3))
# genes
cols_to_mutated_genes <- c(colnames(df.seqz)[c(38:41, 44:54)], "BRCA1/2", "PIK3CA/B")
# pathways
cols_to_mutated_pathways <- colnames(df.seqz)[55:66]

# Exclude PSA
cols_to_features <- c(
  "IsMetastatic", "HasFusion",
  sort(c("Age", "Stage", "GleasonScore",
       "n_SBS", "MATH", "n_INDEL", "Ti_fraction", "Tv_fraction",
       "cnaBurden", "n_CNV", "n_Amp", "n_Del", "ploidy",
       "TD score", 
       "OsCN deletion score",
       "purity"))
)

feature_type <- c(rep("ca", 2L), rep("co", length(cols_to_features) - 2L))
```

### Signatures & features

Analyze the association between signatures and features.

```{r, message=TRUE}
tidy_data.seqz.feature <- get_sig_feature_association(df.seqz,
                                                      cols_to_sigs = cols_to_sigs.seqz,
                                                      cols_to_features = cols_to_features,
                                                      method_co = "pearson",
                                                      type = feature_type, 
                                                      min_n = 2,
                                                      verbose = TRUE) %>%
  get_tidy_association(p_adjust = TRUE)  # adjust p value with FDR

```

> Association of signature exposures with other features was performed using one of two procedures: for a continuous association variable (including ordinal variable), pearson correaltion was performed; for a binary association variable, samples were divided into two groups and Mann-Whitney U-test was performed to test for differences in signature exposure medians between the two groups.

Show the result.

```{r, fig.height=8, fig.width=9}
show_sig_feature_corrplot(tidy_data.seqz.feature, p_val = 0.05)
```

> All shown circles are statistically significant results (FDR < 0.05). A feature will be chopped off from
> the plot if has no correlation with any signatures. Same for the similar plots below.

```{r, include=FALSE}
p = show_sig_feature_corrplot(tidy_data.seqz.feature, p_val = 0.05)
ggsave(filename = "../figures/Figure3A.pdf", plot = p,
       width = 9, height = 8)
```


The result table:

```{r}
DT::datatable(tidy_data.seqz.feature)
```


### Signatures & mutated genes

Analyze the association between signatures and mutated genes (only driver genes identified by MutSig).

```{r, message=TRUE}
tidy_data.seqz.gene <- get_sig_feature_association(df.seqz,
                                                   cols_to_sigs = cols_to_sigs.seqz,
                                                   cols_to_features = cols_to_mutated_genes,
                                                   min_n = 2,
                                                   type = "ca", 
                                                   verbose = TRUE) %>%
  get_tidy_association(p_adjust = TRUE)
```

```{r}
sum_genes = tidy_data.seqz.gene %>% 
  dplyr::filter(p <= 0.05) %>% 
  dplyr::summarise(min = min(measure, na.rm = T),
                   max = max(measure, na.rm = T),
                   sum = list(summary(measure)))

sum_genes$sum
```


```{r, fig.height=5, fig.width=9}
# myPalette <- colorRampPalette(rev(brewer.pal(11, "RdYlBu")))
# sc <- scale_colour_gradientn(colours = myPalette(100), limits=c(-20, 600))

sc <- scale_colour_gradientn(
  colors = c("lightblue", "white", "orange", "red"),
  values = scales::rescale(c(
   sum_genes$min,
    0,
    20,
    sum_genes$max
  )))

show_sig_feature_corrplot(tidy_data.seqz.gene, 
                          ylab = "Mutated genes",
                          ca_gradient_colors = sc,
                          p_val = 0.05) # p_val needs to keep in line with the sum_genes p <= 0.05

```

```{r, include=FALSE}
p = show_sig_feature_corrplot(tidy_data.seqz.gene, 
                          ylab = "Mutated genes",
                          ca_gradient_colors = sc,
                          p_val = 0.05)
ggsave(filename = "../figures/Figure3B.pdf", plot = p,
       width = 8, height = 5)
```


The result table:

```{r}
DT::datatable(tidy_data.seqz.gene)
```

### Signatures & mutated pathways

Analyze the association between signatures and mutated pathways.


```{r, message=TRUE}
tidy_data.seqz.pathways <- get_sig_feature_association(df.seqz,
                                                       cols_to_sigs = cols_to_sigs.seqz,
                                                       cols_to_features = cols_to_mutated_pathways,
                                                       type = "ca", 
                                                       min_n = 2,
                                                       verbose = TRUE) %>%
  get_tidy_association(p_adjust = TRUE)
```


```{r}
sum_pathways = tidy_data.seqz.pathways %>% 
  dplyr::filter(p <= 0.05) %>% 
  dplyr::summarise(min = min(measure, na.rm = T),
                   max = max(measure, na.rm = T),
                   sum = list(summary(measure)))

sum_pathways$sum
```



```{r, fig.height=5, fig.width=9}
sc <- scale_colour_gradientn(
  colors = c("lightblue", "white", "orange", "red"),
  values = scales::rescale(c(
    sum_pathways$min,
    0, 
    20,
    sum_pathways$max
  )))


show_sig_feature_corrplot(tidy_data.seqz.pathways,
                          ylab = "Mutated pathways",
                          ca_gradient_colors = sc,
                          p_val = 0.05)
```


```{r, include=FALSE}
p = show_sig_feature_corrplot(tidy_data.seqz.pathways,
                          ylab = "Mutated pathways",
                          ca_gradient_colors = sc,
                          p_val = 0.05)
ggsave(filename = "../figures/Figure3C.pdf", plot = p,
       width = 9, height = 5)
```

The result table:

```{r}
DT::datatable(tidy_data.seqz.pathways)
```

### Correlation network

To determine the structure of relationship between signatures and features, we use continuous variables to do correlation network analysis.

```{r}
f_dt = df.seqz[, c(cols_to_sigs.seqz, cols_to_features[c(-1, -2)])] %>% 
  dplyr::mutate_if(is.ordered, as.numeric)
```

```{r}
library(corrr)

res.cor <- correlate(f_dt,
  use = "pairwise.complete.obs",
  method = "pearson"
) %>%
  dplyr::mutate_if(is.numeric, dplyr::coalesce, 0)

## Avoid corrr error
## https://github.com/tidymodels/corrr/issues/86
## https://github.com/tidymodels/corrr/issues/34
## https://github.com/tidymodels/corrr/issues/78
set.seed(1234)
res.cor[,-1] = res.cor[, -1] + matrix(
  rnorm(nrow(res.cor[, -1]) * ncol(res.cor[, -1]), mean = 0, sd = 0.01),
  nrow = nrow(res.cor[, -1]),
  byrow = TRUE
)

p = res.cor %>%
  network_plot(min_cor = 0.2,
               colours = rev(c("indianred2", "white", "skyblue1")))

#png("corrr_network.png", width = 7, height = 7, res = 300)
p + scale_size(range=c(0.1, 2))
#ggsave("corrr_network.pdf", plot = p + scale_size(range=c(0.1, 2)))
#dev.off()
```

> variables that are more highly correlated appear closer together and are joined by stronger paths. Paths are also colored by their sign. The proximity of the points are determined using multidimensional clustering.

```{r, include=FALSE}
p = p + scale_size(range=c(0.1, 2))
ggsave(filename = "../figures/corrr_network.pdf", plot = p, width = 7, height = 6)
```



### Signatures & PSA

In all datasets we collected, 3 of them have PSA info. However, the range of PSA value in 3 datasets differs, so I excluded PSA from features above and analyze it independently.

```{r}
df.seqz.psa = df.seqz %>%
  dplyr::filter(!is.na(PSA)) %>%
  dplyr::select(contains("Sig", ignore.case = F), c("PSA", "Study", "CNV_ID"))

df.seqz.psa <- df.seqz.psa %>%
  dplyr:::group_by(Study) %>%
  tidyr::nest() %>%
  dplyr::summarise(
    data = purrr:::map(data, function(x) {
      get_sig_feature_association(x,
                                  cols_to_sigs = cols_to_sigs.seqz,
                                  cols_to_features = "PSA",
                                  method_co = "pearson",
                                  type = "co", verbose = TRUE) %>%
        get_tidy_association(p_adjust = TRUE)
    })
  ) %>%
  tidyr::unnest("data") %>%
  dplyr::filter(feature == "PSA") %>%
  dplyr::mutate(feature = Study)
```

```{r fig.height=5, fig.width=9}
show_sig_feature_corrplot(df.seqz.psa, p_val = 1, breaks_count = c(0, 50, 100, 150, 200), ylab = "PSA")
```

Alought we can observe the statistically significant correlation between signatures and PSA, the results are
inconsistent across different studies. This may be explained by the dynamic property of PSA while not the sample
type due to most of them are primary.

```{r}
table(df.seqz$sample_type, df.seqz$Study)
```

